freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -386,42 +398,60 @@ private boolean usersStateDifferenceChanged(UsersStateDifference beforeSync, Use
 
         return usersStateDifference;
     }
 
     private void applyDifference(UmsUsersState umsUsersState, String environmentCrn, Multimap<String, String> warnings,
-            UsersStateDifference usersStateDifference, FreeIpaClient freeIpaClient, boolean fmsToFreeipaBatchCallEnabled)
-            throws FreeIpaClientException, IOException {
+            UsersStateDifference usersStateDifference, FreeIpaClient freeIpaClient, boolean fmsToFreeipaBatchCallEnabled,
+            boolean credentialsUpdateOptimizationEnabled) throws FreeIpaClientException, IOException {
         LOGGER.debug("Starting {} ...", LogEvent.APPLY_DIFFERENCE_TO_IPA);
         applyStateDifferenceToIpa(environmentCrn, freeIpaClient, usersStateDifference, warnings::put, fmsToFreeipaBatchCallEnabled);
         LOGGER.debug("Finished {}.", LogEvent.APPLY_DIFFERENCE_TO_IPA);
 
         if (!FreeIpaCapabilities.hasSetPasswordHashSupport(freeIpaClient.getConfig())) {
             LOGGER.debug("IPA doesn't have password hash support, no credentials sync required for env:{}", environmentCrn);
         } else {
-            // Sync credentials for all users and not just diff. At present there is no way to identify that there is a change in password for a user
-            LOGGER.debug("Starting {} for {} users ...", LogEvent.SET_WORKLOAD_CREDENTIALS, umsUsersState.getUsersWorkloadCredentialMap().size());
-            workloadCredentialService.setWorkloadCredentials(fmsToFreeipaBatchCallEnabled, freeIpaClient,
-                    umsUsersState.getUsersWorkloadCredentialMap(), warnings::put);
+            // If the credentials update optimization is enabled, sync credentials for users with updated credentials.
+            // Otherwise, sync credentials for all users.
+            LOGGER.debug("Starting {} for {} users ...", LogEvent.SET_WORKLOAD_CREDENTIALS, usersStateDifference.getUsersWithCredentialsToUpdate().size());
+            Map<String, Pair<String, WorkloadCredential>> userToCredentialUpdateParamsMap =
+                    usersStateDifference.getUsersWithCredentialsToUpdate().stream().collect(Collectors.toMap(
+                            Function.identity(), username -> getCredentialUpdateParams(username, umsUsersState)));
+            workloadCredentialService.setWorkloadCredentials(fmsToFreeipaBatchCallEnabled, credentialsUpdateOptimizationEnabled, freeIpaClient,
+                    userToCredentialUpdateParamsMap, warnings::put);
             LOGGER.debug("Finished {}.", LogEvent.SET_WORKLOAD_CREDENTIALS);
         }
     }
 
+    private Pair<String, WorkloadCredential> getCredentialUpdateParams(String username, UmsUsersState umsUsersState) {
+        UserMetadata userMetadata = umsUsersState.getUsersState().getUserMetadataMap().get(username);
+        if (userMetadata == null) {
+            throw new IllegalStateException(String.format("No UserMetadata mapping for user %s", username));
```

Comment:
What happens when this `IllegalStateException` is thrown? Where will it be handled? Won't this stop the whole usersync instead of failing the sync only for one user? Maybe a non runtime exception would be better and force handling these. also some logs would be nice here

Reply:
This is intended as a precondition check, like requireNonNull/checkArgument/etc. If it fails, it means there is a bug. The reason for throwing IllegalStateException, rather than calling Precondition.checkState, is that I saw that being done elsewhere in the Cloudbreak code, and so thought that was the pattern to follow. 

I changed this to use standard requireNonNull checks, which will hopefully clarify the intent.